tiledb/sm/serialization/array_schema.cc

```diff
@@ -304,110 +304,139 @@ Status dimension_to_capnp(
-    tdb_unique_ptr<Dimension>* dimension) {
-  Datatype dim_type = Datatype::ANY;
-  RETURN_NOT_OK(datatype_enum(dimension_reader.getType().cStr(), &dim_type));
-  dimension->reset(tdb_new(Dimension, dimension_reader.getName(), dim_type));
+tuple<Status, optional<ByteVecValue>> tile_extent_from_capnp(
+    const capnp::Dimension::TileExtent::Reader tile_extent_reader,
+    Datatype dim_type) {
+  auto coord_size = datatype_size(dim_type);
+  ByteVecValue tile_extent(coord_size);
+  switch (dim_type) {
+    case Datatype::INT8: {
+      auto val = tile_extent_reader.getInt8();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::UINT8: {
+      auto val = tile_extent_reader.getUint8();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::INT16: {
+      auto val = tile_extent_reader.getInt16();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::UINT16: {
+      auto val = tile_extent_reader.getUint16();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::INT32: {
+      auto val = tile_extent_reader.getInt32();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::UINT32: {
+      auto val = tile_extent_reader.getUint32();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::DATETIME_YEAR:
+    case Datatype::DATETIME_MONTH:
+    case Datatype::DATETIME_WEEK:
+    case Datatype::DATETIME_DAY:
+    case Datatype::DATETIME_HR:
+    case Datatype::DATETIME_MIN:
+    case Datatype::DATETIME_SEC:
+    case Datatype::DATETIME_MS:
+    case Datatype::DATETIME_US:
+    case Datatype::DATETIME_NS:
+    case Datatype::DATETIME_PS:
+    case Datatype::DATETIME_FS:
+    case Datatype::DATETIME_AS:
+    case Datatype::TIME_HR:
+    case Datatype::TIME_MIN:
+    case Datatype::TIME_SEC:
+    case Datatype::TIME_MS:
+    case Datatype::TIME_US:
+    case Datatype::TIME_NS:
+    case Datatype::TIME_PS:
+    case Datatype::TIME_FS:
+    case Datatype::TIME_AS:
+    case Datatype::INT64: {
+      auto val = tile_extent_reader.getInt64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::UINT64: {
+      auto val = tile_extent_reader.getUint64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::FLOAT32: {
+      auto val = tile_extent_reader.getFloat32();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    case Datatype::FLOAT64: {
+      auto val = tile_extent_reader.getFloat64();
+      std::memcpy(tile_extent.data(), &val, coord_size);
+      break;
+    }
+    default:
+      return {LOG_STATUS(Status_SerializationError(
+                  "Error deserializing dimension; unknown datatype.")),
+              nullopt};
```

Comment:
This return statement will move to the check at the top of the function. At that point the `default` branch should never be reach, and this statement should throw `logic_error`.